### PR TITLE
The backend is refactored

### PR DESCRIPTION
**Pull request by eanker**
This pr has an updated version of the backend, which now uses services.

Fixes preview-code/rite-evaluation#240

Together with preview-code/rite-evaluation#277

### CODE REVIEW

File: src/main/java/me/previewcode/backend/DTO/Ordering.java [hunk 1]
```diff
@@ -0,0 +1,26 @@
+package me.previewcode.backend.DTO;
+
+import java.util.List;
+/**
+ * The ordering of the pull request
+ * @author PReview-Code
```
Comment (TimvdLippe) [inline src/main/java/me/previewcode/backend/DTO/Ordering.java]: Let's omit the `@author` tag for all classes.

File: src/main/java/me/previewcode/backend/DTO/Body.java [hunk 2]
```diff
@@ -0,0 +1,15 @@
+package me.previewcode.backend.DTO;
+
+public class Body {
```
Comment (TimvdLippe) [inline src/main/java/me/previewcode/backend/DTO/Body.java]: This name is too generic. Preferably `OrderingInformation`? And maybe we should move this instead into the `Ordering` class since they are tightly coupled.

File: src/main/java/me/previewcode/backend/DTO/Body.java [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+
+        /**
+         * The title of the object
+         */
```
Comment (TimvdLippe) [inline src/main/java/me/previewcode/backend/DTO/Body.java]: The indentation seems off.

File: src/main/java/me/previewcode/backend/DTO/PRGroupComment.java [hunk 4]
```diff
@@ -1,5 +1,14 @@
 package me.previewcode.backend.DTO;
 
+/**
+ * The data for a group comment
+ * 
+ * @author PReview-Code
+ *
+ */
 public class PRGroupComment extends PRComment {
-    public String id;
```
Comment (TimvdLippe) [inline src/main/java/me/previewcode/backend/DTO/PRGroupComment.java]: Why was this renamed?

File: src/main/java/me/previewcode/backend/GithubConnection.java [hunk 5]
```diff
@@ -7,12 +7,31 @@
...
+     * A method that makes a connection with GitHub
+     * 
+     * @param gitHubProvider
+     *            The provider for GitHub data
+     * @throws IOException
+     * 
+     */
     @Inject
-    protected GithubConnection(@Named("github.user") final Provider<GitHub> gitHubProvider) throws IOException {
+    protected GithubConnection(
```
Comment (TimvdLippe) [inline src/main/java/me/previewcode/backend/GithubConnection.java]: I prefer the old formatting.

File: src/main/java/me/previewcode/backend/MainModule.java [hunk 6]
```diff
@@ -38,6 +50,10 @@ public void configureServlets() {
         }
     }
 
+    /**
+     * The method that provides a connection with GitHub
+     * @return the GitHub API instance
```
Comment (TimvdLippe) [inline src/main/java/me/previewcode/backend/MainModule.java]: This method throws an exception instead ;)

`Method to declare Named key "github.user" to obtain the current GitHub instance`

File: src/main/java/me/previewcode/backend/api/v1/CommentsAPI.java [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+import com.google.inject.Inject;
+
+@Path("{owner}/{name}/pulls/{number}/comments/")
+public class CommentsAPI {
+
+    @Inject
+    private GithubCommentsService commentsService;
+
+    /**
+     * A method to set the standard pull request comments
```
Comment (TimvdLippe) [inline src/main/java/me/previewcode/backend/api/v1/CommentsAPI.java]: `A method` is redundant.

File: src/main/java/me/previewcode/backend/services/firebase/FirebaseCommentsService.java [hunk 8]
```diff
@@ -0,0 +1,20 @@
...
+import me.previewcode.backend.FirebaseConnection;
+
+import com.google.firebase.database.DatabaseReference;
+import com.google.inject.Singleton;
+
+@Singleton
+public class FirebaseCommentsService extends FirebaseConnection {
+    DatabaseReference projectRef = this.ref;
+
+    public void setComments(String owner, String name, int number,
```
Comment (TimvdLippe) [inline src/main/java/me/previewcode/backend/services/firebase/FirebaseCommentsService.java]: No documentation ;)

File: src/main/java/me/previewcode/backend/DTO/Body.java (hunk 2 above)
Comment (eanker) [inline src/main/java/me/previewcode/backend/DTO/Body.java]: Well, I extracted this one so that PRBody extends this class.

Because they both have title and description.

Should I make to separate classes?

File: src/main/java/me/previewcode/backend/DTO/PRGroupComment.java (hunk 4 above)
Comment (eanker) [inline src/main/java/me/previewcode/backend/DTO/PRGroupComment.java]: I thought id might be ambigous, because it could stand for comment or group ID.

File: src/main/java/me/previewcode/backend/services/firebase/FirebaseCommentsService.java [hunk 9]
```diff
@@ -0,0 +1,20 @@
...
+            Integer commentID, String groupID) {
+        owner = owner.toLowerCase();
+        name = name.toLowerCase();
+        projectRef.child(owner).child(name).child("pulls")
```
Comment (TimvdLippe) [inline src/main/java/me/previewcode/backend/services/firebase/FirebaseCommentsService.java]: Should be `this.ref` and the field should be removed.

File: src/main/java/me/previewcode/backend/services/github/GithubCommentsService.java [hunk 10]
```diff
@@ -0,0 +1,110 @@
...
+
+@Singleton
+public class GithubCommentsService extends GithubConnection {
+    private GHRepository repo;
+
+    @Inject
+    private FirebaseCommentsService pullService;
+
+    @Inject
+    public GithubCommentsService(
```
Comment (TimvdLippe) [inline src/main/java/me/previewcode/backend/services/github/GithubCommentsService.java]: I would put the argument on the same line.

File: src/main/java/me/previewcode/backend/services/github/GithubCommentsService.java [hunk 11]
```diff
@@ -0,0 +1,110 @@
...
+
+    private GHIssueComment postComment(String owner, String name, int number,
+            PRComment comment) {
+        try {
+            repo = this.githubProvider.get().getRepository(
+                    owner.toLowerCase() + "/" + name.toLowerCase());
+            GHPullRequest pr = repo.getPullRequest(number);
+            return pr.comment(comment.body);
+        } catch (IOException e) {
+            e.printStackTrace();
```
Comment (TimvdLippe) [inline src/main/java/me/previewcode/backend/services/github/GithubCommentsService.java]: Printing the stacktrace should not be necessary since we are immediately rethrowing.

File: src/main/java/me/previewcode/backend/services/github/PullRequestService.java [hunk 12]
```diff
@@ -0,0 +1,64 @@
...
+     * @param name
+     *            The owner of the repository on which the pull request is created
+     * @param body
+     *            The body of the pull request
+     * @return The number of the newly made pull request
+     */
+    public PrNumber createPullRequest(String owner, String name, PRbody body) {
+        PrNumber number = new PrNumber();
+        try {
+            repo = this.githubProvider.get().getRepository(
```
Comment (TimvdLippe) [inline src/main/java/me/previewcode/backend/services/github/PullRequestService.java]: This can be a local variable.

File: src/main/java/me/previewcode/backend/DTO/PRGroupComment.java (hunk 4 above)
Comment (TimvdLippe) [inline src/main/java/me/previewcode/backend/DTO/PRGroupComment.java]: Imo `id` is just fine. WDYT @txsmith ?

**Comment by eanker** [comment]
Updated the pr according to the changes.

File: src/main/java/me/previewcode/backend/services/github/GithubCommentsService.java [hunk 13]
```diff
@@ -0,0 +1,110 @@
...
+        pullService.setComments(owner, name, number, newComment.getId(),
+                comment.groupID);
+
+        return setResponseComment(newComment);
+    }
+
+    private GHIssueComment postComment(String owner, String name, int number,
+            PRComment comment) {
+        try {
+            repo = this.githubProvider.get().getRepository(
```
Comment (TimvdLippe) [inline src/main/java/me/previewcode/backend/services/github/GithubCommentsService.java]: This should be a local variable. Since this is a singleton, it is possible that concurrent requests enter this method and therefore post comments on the wrong repo.

File: src/main/java/me/previewcode/backend/services/github/GithubCommentsService.java [hunk 14]
```diff
@@ -0,0 +1,110 @@
...
+                    owner.toLowerCase() + "/" + name.toLowerCase());
+            GHPullRequest pr = repo.getPullRequest(number);
+            return pr.comment(comment.body);
+        } catch (IOException e) {
+            throw new IllegalArgumentException(e);
+        }
+    }
+
+    private PRresponseComment setResponseComment(GHIssueComment comment) {
+        PRresponseComment responseComment = new PRresponseComment();
```
Comment (TimvdLippe) [inline src/main/java/me/previewcode/backend/services/github/GithubCommentsService.java]: This is something for the API to figure out, not what a service should do.

File: src/main/java/me/previewcode/backend/services/github/PullRequestService.java [hunk 15]
```diff
@@ -0,0 +1,67 @@
...
+
+/**
+ * The service for creating a pull request on GitHub
+ *
+ */
+@Singleton
+public class PullRequestService extends GithubConnection {
+
+    @Inject
+    StatusService statusSerivce;
```
Comment (TimvdLippe) [inline src/main/java/me/previewcode/backend/services/github/PullRequestService.java]: I think instead of this service knowing what a different service does, the API should call the statusservice instead.

**Comment by TimvdLippe** [comment]
LGTM!
